torch/csrc/cuda/nccl.cpp

```diff
@@ -123,6 +136,57 @@ static inline void NCCL_CHECK(ncclResult_t result) {
   NCCL_CHECK(from_nccl_result(result));
 }
 
+static inline void NCCL_CHECK_NONBLOCKING(ncclResult status, ncclComm_t comm) {
+#ifdef NCCL_HAS_COMM_NONBLOCKING
+  ncclResult_t result = to_nccl_result(status);
+  while (result == ncclInProgress) {
+    ncclCommGetAsyncError(to_nccl_comm(comm), &result);
+  }
+  if (result != ncclSuccess) {
+    throw_nccl_error(from_nccl_result(result));
+  }
+#else
+  TORCH_INTERNAL_ASSERT(
+      false, "NCCL COMM NONBLOCKING USED WITH UNSUPPORTED NCCL VERSION.");
+#endif
+}
+
+static inline void NCCL_CHECK_NONBLOCKING(
+    ncclResult_t result,
+    ncclComm_t comm) {
+  NCCL_CHECK_NONBLOCKING(from_nccl_result(result), comm);
+}
+
+static inline void NCCL_CHECK_NONBLOCKING_GROUPEND(
+    ncclResult status,
+    std::vector<ncclComm_t>& comms) {
+#ifdef NCCL_HAS_COMM_NONBLOCKING
+  ncclResult_t result = to_nccl_result(status);
+  if (result == ncclInProgress) {
+    for (const auto i : c10::irange(comms.size())) {
+      do {
+        ncclCommGetAsyncError(to_nccl_comm(comms[i]), &result);
+      } while (result == ncclInProgress);
```

Comment:
I wonder which one should be the inner loop. 

Would it be possible that a device is hanging, while another already errors out, in which case we would miss catching the error here?